python/tank/authentication/user.py

```diff
@@ -118,6 +128,109 @@ def impl(self):
         return self._impl
 
 
+class ShotgunSamlUser(ShotgunUser):
+    """
+    This specialized shotgun user is needed when SSO is used, as it provides
+    mechanisms for automatic claims renewal.
+
+    User objects are created via the :class:`ShotgunAuthenticator` object, which will handle
+    caching user objects on disk, prompting the user for their credentials etc.
+
+    This specialized class allows the user to query the claims expiration and
+    see if `interactive_authentication.renew_session(user)` needs to be called.
+
+    It is also possible to start/stop/query the state of the automatic claims renewal:
+
+        user = ...
+        if isinstance(user, ShotgunSamlUser):
+            user.start_claims_renewal()
+        ...
+        if user.is_claims_renewal_active():
+            user.stop_claims_renewal()
+    """
+
+    def __init__(self, impl):
+        """
+        :param impl: Internal user implementation class this class proxies.
+        """
+        super(ShotgunSamlUser, self).__init__(impl)
+        self._timer = None
+
+    def get_claims_expiration(self):
+        """
+        Obtain claims expiration for the user.
+
+        :returns: The expiration in seconds since January 1st 1970 UTC.
+        """
+        # The import must be located here. If it is at global scope, there
+        # is an import error.
+        from .shotgun_shared import saml2_sso
+
+        return saml2_sso.get_saml_claims_expiration(self._impl.get_cookies())
+
+    def _automatic_claims_renewal(self, preemtive_renewal_threshold=0.9):
+        """
+        Handles automatic renewal of the SAML2 claims for the user.
+
+        :params user: an already logged in user.
+        :params preemtive_renewal_threshold: How far into the claims duration we will attempt renewal.
+                                             Defaults to 90%, usually 3 minutes 45 seconds (90% of 5 mins).
+        """
+        logger.debug("Automatic claims renewal")
```

Comment:
can we make this more informative? Imagine a tech support reading these logs. Perhaps something like `Starting to perform automatic SSO claim renewal`